## Darshan Patel

# 1.	Use “heart.csv” dataset 

### a.	Study the dataset, identify the independent and target variables. 

In [270]:
# Importing all necessary libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [271]:
df = pd.read_csv("C:/Users/Darshan/Desktop/SEM-2 AI and DS/ML/heart.csv")
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [272]:
df.shape

(303, 14)

In [273]:
df['target'].unique()

array([1, 0], dtype=int64)

#### We can see dataset and it has total 303 rows and 14 columns. 
#### Target Attribute is target attribute which contain 2 different values 0 and 1.
#### 0 --> Not suffered
#### 1 --> Suffered

#### rest of the attributes are Independent attributes.

In [274]:
# Check for null or missing values in our data
df.isnull().sum()

# We can see that there is no null or missing values in our data

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [275]:
# Check for statistical measures 
df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [276]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


### b.	Break down the dataset into training (70%) and testing (30%) datasets and make sure these two datasets do not share any common data point.

In [277]:
# Assign dependent and indepedent data to 2 different variables

X = df.drop(columns = 'target')
y = df['target']
print(X.shape)
print(y.shape)

(303, 13)
(303,)


In [278]:
# Lets split data into training and testing set using sklearn

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

#### This code divide data into  training (70%) and testing (30%) groups and we do not need to worry about data overlapping as "train_test_split" library take care of it by default.

### c.	Then, fit a decision tree classifier on the training data using the target and independent variables.

In [279]:
from sklearn.tree import DecisionTreeClassifier

In [280]:
model = DecisionTreeClassifier(criterion = 'entropy', random_state = 0) # Here we used Decision Tree as a classifier
model.fit(X_train, y_train)    # Fitted data for training using training set

DecisionTreeClassifier(criterion='entropy', random_state=0)

### d.	Predict the target variable on the testing dataset.

In [281]:
y_pred = model.predict(X_test)    # making predictions on test data

In [282]:
y_pred

array([0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 1], dtype=int64)

In [283]:
result = pd.DataFrame({"Actual":y_test,"Predicted":y_pred})
result.head(10)

Actual  Predicted
179       0          0
228       0          1
111       1          1
246       0          0
60        1          1
9         1          1
119       1          1
223       0          0
268       0          0
33        1          0

### e.	Comment on the performance of this classifier on training vs. testing datasets.

In [284]:
# applying Confusion Matrix to evaluate the performance of our model

from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix \n",cm)

Confusion Matrix 
 [[31 10]
 [15 35]]


In [285]:
# Checking for training and testing Accuracy
train1 = round(model.score(X_train,y_train)*100,2)
test1 = round(model.score(X_test,y_test)*100,2)

print("Training-set score:" ,train1)
print("Test-set score: " ,test1)

Training-set score: 100.0
Test-set score:  72.53


In [286]:
# checking Accuracy of model

from sklearn.metrics import accuracy_score
accuracy1 =accuracy_score(y_test,y_pred)*100
print("Accuracy of model is: ",accuracy*100)

Accuracy of model is:  72.52747252747253


In [287]:
# Checking for Precision and Recall

from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

precision = precision_score(y_test,y_pred)
recall = recall_score(y_test,y_pred)

print("Precision score is :",precision)
print("Recall score is :",recall)

Precision score is : 0.7777777777777778
Recall score is : 0.7


In [288]:
# Checking F1 score of model

from sklearn.metrics import f1_score
f1 = f1_score(y_test,y_pred)
print("F-1 score is :",f1)

F-1 score is : 0.7368421052631577


# 2.	Now follow these steps:

### a.	Break down the same dataset into training (70%) and testing (30%) datasets as you did in Step (1).

In [289]:
# Lets split data into training and testing set using sklearn

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((212, 13), (91, 13), (212,), (91,))

#### This code divide data into  training (70%) and testing (30%) groups and we do not need to worry about data overlapping as "train_test_split" library take care of it by default.

### b.	Apply 10-Fold Cross Validation on whole dataset

In [290]:
from sklearn.model_selection import KFold, cross_val_score, cross_val_predict
kf = KFold(n_splits=10)                                       # Apply 10-Fold Cross Validation

cnt = 1
for train_index, test_index in kf.split(X, y):
    print(f'Fold:{cnt}, Train set: {len(train_index)}, Test set:{len(test_index)}')
    cnt += 1

Fold:1, Train set: 272, Test set:31
Fold:2, Train set: 272, Test set:31
Fold:3, Train set: 272, Test set:31
Fold:4, Train set: 273, Test set:30
Fold:5, Train set: 273, Test set:30
Fold:6, Train set: 273, Test set:30
Fold:7, Train set: 273, Test set:30
Fold:8, Train set: 273, Test set:30
Fold:9, Train set: 273, Test set:30
Fold:10, Train set: 273, Test set:30


In [291]:
# In above code we successfully applied 10 fold cros validation which generate 10 sets for training and testing

### c.  Then, use the same classifier as (1) on the training data. 

In [292]:
model = DecisionTreeClassifier(criterion = "entropy")
training_scores = cross_val_score(model, X_train, y_train, cv=kf)
training_scores

array([0.81818182, 0.68181818, 0.61904762, 0.76190476, 0.71428571,
       0.76190476, 0.71428571, 0.9047619 , 0.80952381, 0.66666667])

In [293]:
# Here we used same classifiers as before with 10 Fold cross validation

In [294]:
score.mean()     # mean Accuracy on train data

0.7387096774193548

In [295]:
model.fit(X_train, y_train)
model.score(X_train, y_train)   # Train score on 10 fold

1.0

In [296]:
model_pred = cross_val_predict(model, X_train, y_train, cv=kf)
accuracy2 = accuracy_score(y_train, model_pred)*100

accuracy2
# Prediction on training dataset

74.05660377358491

### d.	Predict the target variable using the testing dataset.

In [297]:
from sklearn.metrics import accuracy_score

prediction = cross_val_predict(model, X_test, y_test, cv=kf)
test2 = accuracy_score(y_test, prediction)
test2*100

81.31868131868131

In [298]:
# Making prediction on test dataset after applying 10 fold cross validation

In [299]:
result = pd.DataFrame({"Actual":y_test,"Predicted":prediction})
result

Actual  Predicted
179       0          1
228       0          0
111       1          1
246       0          0
60        1          1
..      ...        ...
250       0          0
19        1          1
143       1          1
79        1          1
144       1          1

[91 rows x 2 columns]

### e.	Comment on the performance of the classifier on training vs. testing datasets. 

In [300]:
# In the previsous model we got accuracy of 72.52 % and after applying 10 fold cross validation we got 90.4% of highest
# Accuracy in 2nd fold where as lowest is 61.9 % in 10th fold.

# After 10 fold cross validation average accuracy of model is : 74.72 % which is slightly better than previous model

### f.	Does the performance change from Step (1) to Step (2)? Can you explain the change? Tip: you may think about cross validation and if it affects overfitting. 


In [301]:
# yes, after using 10 fold cross validation accuracy is increased slightly.
# after 10 fold it eliminates the ptoblem of overfitting by dividing data into 10 train and test sets 
# and finding accuracy for each set and than we find mean accuracy  

# 3.	Now follow these steps:

### a.	Change the depth of decision tree classifier from Step (2). If you are using  DecisionTreeClassifier in Pyhton, you can change max_depth parameter to change the depth of the tree. Try the following values for max_depth: 2, 5, 10, 12. 

In [302]:
## we will apply gridsearch algorithm to findout best fitted depth for our model.


In [303]:

from sklearn.model_selection import GridSearchCV

grid_parameter = {'criterion' : ['gini','entropy'],'max_depth' : [2,5,10,12],'min_samples_leaf' : range(1,10,1),
    'min_samples_split' : range(2,10,1),'splitter' : ['best','random'] }

grid_search = GridSearchCV(model,param_grid=grid_parameter,cv=10,n_jobs=-1)

In [304]:
# Fitting training data using grid search

grid_search.fit(X_train,y_train)

GridSearchCV(cv=10, estimator=DecisionTreeClassifier(criterion='entropy'),
             n_jobs=-1,
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [2, 5, 10, 12],
                         'min_samples_leaf': range(1, 10),
                         'min_samples_split': range(2, 10),
                         'splitter': ['best', 'random']})

In [305]:
parameters_best = grid_search.best_params_   # it will return best parameters for model

print(parameters_best)

{'criterion': 'gini', 'max_depth': 12, 'min_samples_leaf': 5, 'min_samples_split': 5, 'splitter': 'random'}


In [306]:
print("Score: ", grid_search.best_score_)

Score:  0.8538961038961039


In [307]:
# We can see that the method 'best_score_' returen max depth as 5 which is best suited for our model from all of the listed
# depths.

### b.	Comment on the performance of the classifier on the training vs. testing datasets. 

In [308]:
model = DecisionTreeClassifier(criterion = 'gini', max_depth =  5, min_samples_leaf  = 8, min_samples_split = 7, splitter= 'random')
model.fit(X_train,y_train)

DecisionTreeClassifier(max_depth=5, min_samples_leaf=8, min_samples_split=7,
                       splitter='random')

In [309]:
train3 = model.score(X_train,y_train)
train3

0.8207547169811321

In [310]:
model.score(X_test,y_test)*100

78.02197802197803

In [311]:
y_pred =model.predict(X_test)
y_pred

array([0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1,
       1, 0, 1], dtype=int64)

In [312]:
print("Training score: ", model.score(X_train, y_train)*100)

print("Test score: ", model.score(X_test, y_test)*100)

Training score:  82.0754716981132
Test score:  78.02197802197803


In [313]:
from sklearn.metrics import accuracy_score
accuracy3 = accuracy_score(y_test,y_pred)

In [314]:
# We can see that classifier give 83% accuracy on training dataset whereas 
# 73% on testing dataset

### c  Which max_depth value would you set based on your observations in (b)? Why?

### d What do you conclude in terms of tree depth, performance and overfitting based on (b)?

# 4.	Now follow these steps:

### a.	Try Random Forest on the training dataset

In [315]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=355)

In [316]:
from sklearn.ensemble import RandomForestClassifier
forest = RandomForestClassifier()

In [317]:
forest.fit(X_train,y_train)

RandomForestClassifier()

### b.	Predict the target variable using the testing dataset.

In [318]:
y_pred = forest.predict(X_test)
y_pred

array([1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 1], dtype=int64)

### c.	Comment on the performance of the classifier on training vs. testing datasets.

In [319]:
from sklearn.metrics import accuracy_score
accuracy4 = round((accuracy_score(y_test, y_pred)*100),2)
print("Random Forest accuracy score: ", accuracy_score(y_test, y_pred)*100)

Random Forest accuracy score:  81.31868131868131


In [320]:
print("Random forest training score :" ,forest.score(X_train, y_train)*100)

print("Random forest test score :",forest.score(X_test, y_test)*100)

Random forest training score : 100.0
Random forest test score : 81.31868131868131


In [321]:
cm = confusion_matrix(y_test,y_pred)
cm

array([[28,  9],
       [ 8, 46]], dtype=int64)

### d.	Does the performance change from Step (2) to Step (4)? Can you explain the change? Tip: you may create a table summarizing/comparing the performance in Step (1), (2), (3) and (4).

In [322]:
df = pd.DataFrame({"Decision_Tree":{"Accuracy": accuracy1},"Decision_Tree(10-Fold)":{"Accuracy": accuracy2},
                   "Decision_Tree(Max_depth)":{"Accuracy":accuracy3*100},"RandomForest":{"Accuracy":accuracy4}})
df

Decision_Tree  Decision_Tree(10-Fold)  Decision_Tree(Max_depth)  \
Accuracy      72.527473               74.056604                 78.021978   

          RandomForest  
Accuracy         81.32